File: packages/typescript/infermap/src/core/detect.ts
// Lightweight domain-pack auto-detection from column names.
import { listDomains, loadDomain } from "goldencheck-types";
import type { DetectionResult } from "goldencheck-types";
import { getInfermapBackend } from "./wasm/backend.js";

export const DEFAULT_MIN_SCORE = 0.3;

const TOKEN_SPLIT = /[_\-.\s]+/;

export interface DetectInput {
  columns: string[];
}

function tokens(s: string): string[] {
  return s.toLowerCase().split(TOKEN_SPLIT).filter(Boolean);
}

/** True iff `hint`'s tokens appear as a contiguous run in `col`'s tokens.
 *  Replaces the prior symmetric `c.includes(h) || h.includes(c)`, which
 *  fired on partial overlaps — a 2-char hint like `"id"` matched
 *  `account_id`, `paid`, `void_id` indiscriminately. Token-boundary
 *  matching keeps `"npi"` matching `provider_npi` while rejecting it
 *  against `"npiece"`. */
function hintMatches(hint: string, col: string): boolean {
  const h = tokens(hint);
  const c = tokens(col);
  if (h.length === 0 || c.length === 0) return false;
  for (let i = 0; i <= c.length - h.length; i++) {
    let ok = true;
    for (let j = 0; j < h.length; j++) {
      if (c[i + j] !== h[j]) {
        ok = false;
        break;
      }
    }
    if (ok) return true;
  }
  return false;
}

/** Pure scoring over resolved [name, hints[]] domains — the WASM parity oracle.
 *  Byte-identical to infermap-core::detect_domain for NON-EMPTY columns. The
 *  empty-columns `no_data` guard lives in detectDomainDetailed (the kernel guards
 *  it too), so scoreDomains is never called with empty columns in production. */
export function scoreDomains(
  columns: string[],
  resolved: Array<[string, string[]]>,
  minScore: number,
): DetectionResult {
  const scored: Array<[string, number]> = [];
  for (const [name, hints] of resolved) {
    if (hints.length === 0) continue; // == Rust `hints.is_empty()` skip
    let hits = 0;
    for (const c of columns) {
      for (const h of hints) {
        if (hintMatches(h, c)) {
          hits++;
          break;
        }
      }
    }
    scored.push([name, hits / Math.max(columns.length, 1)]);
  }
  if (scored.length === 0) {
    return { domain: null, score: 0, runner_up: null, runner_up_score: 0, reason: "no_data" };
  }
  scored.sort((a, b) => b[1] - a[1]);
  const [bestName, bestScore] = scored[0]!;
  const runner = scored[1];
  const runnerName = runner ? runner[0] : null;
  const runnerScore = runner ? runner[1] : 0;
  if (bestScore < minScore) {
    return { domain: null, score: bestScore, runner_up: runnerName, runner_up_score: runnerScore, reason: "below_min_score" };
  }
  const topCount = scored.filter(([, s]) => s === bestScore).length;
  if (topCount > 1) {
    return { domain: null, score: bestScore, runner_up: runnerName, runner_up_score: runnerScore, reason: "tie" };
  }
  return { domain: bestName, score: bestScore, runner_up: runnerName, runner_up_score: runnerScore, reason: "confident" };
}

export function detectDomain(
  input: DetectInput | { records?: ReadonlyArray<Record<string, unknown>> },
  candidates?: string[],
  minScore: number = DEFAULT_MIN_SCORE,
): string | null {
  return detectDomainDetailed(input, candidates, minScore).domain;
}

/** Auto-detect with full diagnostic info. See `DetectionResult` for the
 *  shape. Callers like `goldenpipe`'s infer_schema stage use this to
 *  distinguish "confident pick" from "tied" / "below threshold" /
 *  "no data" and surface that to the InferredSchema's evidence map. */
export function detectDomainDetailed(
  input: DetectInput | { records?: ReadonlyArray<Record<string, unknown>> },
  candidates?: string[],
  minScore: number = DEFAULT_MIN_SCORE,
): DetectionResult {
  let columns: string[];
  if ("columns" in input && Array.isArray(input.columns)) {
    columns = input.columns;
  } else if ("records" in input && input.records && input.records.length > 0) {
    columns = Object.keys(input.records[0]!);
  } else {
    return { domain: null, score: 0, runner_up: null, runner_up_score: 0, reason: "no_data" };
  }
  if (columns.length === 0) {
    return { domain: null, score: 0, runner_up: null, runner_up_score: 0, reason: "no_data" };
  }

  const domainNames = candidates ?? listDomains().filter((d) => d !== "generic");
  // Hoist hint resolution into a pre-pass so the SAME resolved input feeds the
  // kernel or the pure path. Empty-hint domains are INCLUDED (both paths skip
  // them: kernel via hints.is_empty(), pure via hints.length===0), so scoring is
  // identical to the pre-refactor inline loop.
  const resolved: Array<[string, string[]]> = domainNames.map((d) => {
    const pack = loadDomain(d);
    const allHints = new Set<string>();
    for (const spec of Object.values(pack.types)) {
      for (const h of spec.name_hints) allHints.add(h);
    }
    return [d, Array.from(allHints)];
  });

  const backend = getInfermapBackend();
  if (backend) {
    return backend.detectDomain(columns, resolved, minScore);
  }
  return scoreDomains(columns, resolved, minScore);
}

File: packages/python/infermap/infermap/detect.py
"""Lightweight domain-pack auto-detection from column names."""
from __future__ import annotations

import re

from goldencheck_types import DetectionResult, list_domains, load_domain

from infermap._native_loader import native_enabled, native_module

DEFAULT_MIN_SCORE = 0.3

# Split column / hint identifiers on the usual word separators. Keeping the
# split conservative — `_-.` plus whitespace — matches how SQL columns,
# CSV headers, and Python identifiers are conventionally tokenized.
_TOKEN_SPLIT = re.compile(r"[_\-.\s]+")


def _tokens(s: str) -> list[str]:
    return [t for t in _TOKEN_SPLIT.split(s.lower()) if t]


def _hint_matches(hint: str, col: str) -> bool:
    """True iff ``hint``'s tokens appear as a contiguous run in ``col``'s tokens.

    Replaces the prior symmetric substring check ``h in c or c in h``, which
    fired on partial overlaps — a 2-char hint like ``"id"`` matched
    ``account_id``, ``paid``, ``void_id``, ``id_card`` indiscriminately and
    the reverse direction let any short column match every long hint.
    Token-boundary matching keeps ``"npi"`` matching ``provider_npi`` (token
    membership) while rejecting ``"npi"`` against ``"npiece"`` (no boundary).
    """
    h_tokens = _tokens(hint)
    c_tokens = _tokens(col)
    if not h_tokens or not c_tokens:
        return False
    n = len(h_tokens)
    return any(c_tokens[i : i + n] == h_tokens for i in range(len(c_tokens) - n + 1))


def detect_domain(
    df,
    candidates: list[str] | None = None,
    min_score: float = DEFAULT_MIN_SCORE,
) -> str | None:
    """Pick the domain pack whose name_hints best match this df's columns.

    Returns ``None`` if no candidate scores at or above ``min_score``, or if
    multiple candidates tie for top score. Use ``detect_domain_detailed``
    when you want to distinguish those two cases or see the runner-up.
    The ``generic`` pack is always excluded from auto-detection.
    """
    return detect_domain_detailed(df, candidates, min_score).domain


def detect_domain_detailed(
    df,
    candidates: list[str] | None = None,
    min_score: float = DEFAULT_MIN_SCORE,
) -> DetectionResult:
    """Auto-detect with full diagnostic info.

    Returns a :class:`DetectionResult` carrying:
      - ``domain`` — the picked name, or None when no decision was made
      - ``score`` — top score (regardless of whether it was picked)
      - ``runner_up`` / ``runner_up_score`` — second-best candidate
      - ``reason`` — one of "confident", "tie", "below_min_score", "no_data"

    Callers like ``goldenpipe.stages.infer_schema`` use this to surface
    "auto-detected with confidence" vs "auto-fallback because tied" in
    the InferredSchema's confidence field and evidence map.
    """
    columns = [str(c) for c in df.columns]
    # Load packs (host / "smart pipe"): (name, deduped name_hints) IN iteration order.
    # The scoring + decision ("dumb kernel") is single-sourced via `_detect_core`.
    domain_names = candidates or [d for d in list_domains() if d != "generic"]
    domains: list[tuple[str, list[str]]] = []
    for d in domain_names:
        pack = load_domain(d)
        hints = list({h for spec in pack.types.values() for h in spec.name_hints})
        domains.append((d, hints))

    domain, score, runner_up, runner_up_score, reason = _detect_core(columns, domains, min_score)
    return DetectionResult(
        domain=domain, score=score,
        runner_up=runner_up, runner_up_score=runner_up_score,
        reason=reason,
    )


def _detect_core(
    columns: list[str],
    domains: list[tuple[str, list[str]]],
    min_score: float,
) -> tuple[str | None, float, str | None, float, str]:
    """Dispatch the detect scoring+decision to the native kernel when gated, else pure."""
    if native_enabled("detect_domain"):
        return tuple(native_module().detect_domain(columns, domains, min_score))
    return _detect_core_pure(columns, domains, min_score)


def _detect_core_pure(
    columns: list[str],
    domains: list[tuple[str, list[str]]],
    min_score: float,
) -> tuple[str | None, float, str | None, float, str]:
    """Byte-identical reference for ``infermap-core::detect_domain``.

    Returns ``(domain, score, runner_up, runner_up_score, reason)``. Empty columns or
    no scorable domain (all hint-less) => ``no_data``. Direction-aware sort is stable
    (Python ``sort(reverse=True)`` keeps ties in host order).
    """
    if not columns:
        return (None, 0.0, None, 0.0, "no_data")
    scored: list[tuple[str, float]] = []
    for name, hints in domains:
        if not hints:
            continue
        hits = sum(1 for c in columns if any(_hint_matches(h, c) for h in hints))
        scored.append((name, hits / len(columns)))
    if not scored:
        return (None, 0.0, None, 0.0, "no_data")
    scored.sort(key=lambda x: x[1], reverse=True)
    best_name, best_score = scored[0]
    runner_name, runner_score = (scored[1] if len(scored) > 1 else (None, 0.0))
    if best_score < min_score:
        return (None, best_score, runner_name, runner_score, "below_min_score")
    if sum(1 for _, s in scored if s == best_score) > 1:
        return (None, best_score, runner_name, runner_score, "tie")
    return (best_name, best_score, runner_name, runner_score, "confident")
